db/db_impl/db_impl_files.cc

```diff
@@ -892,20 +892,28 @@ Status DBImpl::SetDBId(bool read_only) {
     if (immutable_db_options_.write_dbid_to_manifest && s.ok()) {
       VersionEdit edit;
       edit.SetDBId(db_id_);
-      Options options;
-      MutableCFOptions mutable_cf_options(options);
       versions_->db_id_ = db_id_;
-      s = versions_->LogAndApply(versions_->GetColumnFamilySet()->GetDefault(),
-                                 mutable_cf_options, &edit, &mutex_, nullptr,
-                                 /* new_descriptor_log */ false);
+      if (version_edits_ctx == nullptr) {
+        // In case of ReadOnly DB, recovery_version_edits is nullptr.
```

Comment:
https://github.com/facebook/rocksdb/blob/627deb7ceb732d62df81e283ed425119fadca0ff/db/db_basic_test.cc#L185. This test specifically tests `write_dbid_to_manifest` with `read_only` because of which I added these if else checks. Am I missing something?

Reply:
Good finding. I think this test was added by mistake. We intentionally pay attention not to create or write to the files if DB is opened for read-only access. One example is https://github.com/facebook/rocksdb/blob/7.1.fb/db/db_impl/db_impl_open.cc#L57. If db is for read-only, we don't even write LOG files. I also remember we have use cases in which application has only read access to the file system.
cc @siying who has more context on the `write_dbid_to_manifest` feature, and @ajkr and @pdillinger who, iirc, have been involved in the discussion of users of `DBImplReadOnly` having only read access to the file system.

Reply:
I'm not really following the details here, but btw there is a ReadOnlyFileSystem that can be used in testing read-only DB to check that we don't write to FS.

Reply:
@pdillinger  The above unit test (in comments above) opens`readonly` db and set the `write_dbid_to_manifest` option to true. So I wanted to confirm do we write db_id to manifest in case of readonly? If not we should remove the unit test.

Reply:
This is the issue #7966 with more discussions on how we should handle `DBImplReadOnly`.

Reply:
`write_dbid_to_manifest = true` is not required for the test to pass anyway, so I consider it a low quality test. And I definitely agree that setting `write_dbid_to_manifest = true` should not supersede "read only" open. Ideally we would test that the combination does not actually write to FS, so perhaps the existing (low quality) test can be updated for that (e.g. open read-only with CompositeEnvWrapper+ReadOnlyFileSystem).

Reply:
As @riversand963 mentioned,  should we return an error in `ValidateOptions()` if `read_only` and `write_dbid_to_manifest` are both true?

Reply:
I don't think we need to. (a) It was envisioned that write_dbid_to_manifest=true would be eventual default, so such a limitation would need to be changed if we did that. (b) I think of "read only" as amending the Options, so that you can use the same Options from read-write to open read-only.